maa-cli/src/run/copilot.rs

```diff
@@ -0,0 +1,204 @@
+use super::run;
+use crate::{
+    config::task::{
+        default_variants, task_type::TaskType, value::input::Input, Strategy, Task, TaskList, Value,
+    },
+    dirs::Dirs,
+    installer::maa_core::find_resource,
+    normal, object, warning,
+};
+use anyhow::anyhow;
+use anyhow::Context;
+use anyhow::Result;
+use prettytable::{format, row, Table};
+use reqwest::blocking::Client;
+use serde_json::Value as JsonValue;
+use std::{
+    fs::{self, File},
+    io::{self, BufReader, Write},
+    path::{Path, PathBuf},
+};
+
+pub fn copilot(
+    dirs: &Dirs,
+    link: String,
+    addr: Option<String>,
+    user_resource: bool,
+    batch: bool,
+    dry_run: bool,
+) -> Result<()> {
+    let api = "https://prts.maa.plus/copilot/get/";
+    let cache_dir = dirs.cache().display().to_string();
+    let mut stage_dir = match find_resource(dirs).context("Failed to find resource!") {
+        Ok(stage_dir) => stage_dir,
+        Err(_) => return Err(anyhow!("Resource not found")),
+    };
+    stage_dir.push("Arknights-Tile-Pos");
+
+    let copilot_id = match link.get(link.len() - 5..) {
+        Some(id) => id,
+        None => {
+            return Err(anyhow!("Code Invaild"));
+        }
+    };
+
+    let url = api.to_owned() + copilot_id;
+
+    // Visit the API to obtain the data of homework.
+    let client = Client::new();
+    let response = client.get(&url).send().unwrap();
+    let json: JsonValue = response.json().unwrap();
+    let status_code = json["status_code"].clone().to_string();
+    if status_code != "200" {
+        return Err(anyhow!("Request Error"));
+    }
+
+    // Extract the response
+    let context = json["data"]["content"].clone();
+    let content = match context.as_str() {
+        Some(content) => content.to_string(),
+        None => {
+            return Err(anyhow!("JSON Prase Error"));
+        }
+    };
+
+    let recast: JsonValue = serde_json::from_str(&content).unwrap();
+
+    // Determine type of stage
+    let mut task_type: &str = "Copilot";
+    let content_type = recast["type"].clone();
+    if content_type == "SSS" {
+        task_type = "SSSCopilot";
+    }
+
+    // Print stage and operators info
+    if task_type == "Copilot" {
+        let stage_code_name = recast["stage_name"].as_str().unwrap();
+        let stage_name = match find_stage_name(stage_code_name, stage_dir) {
+            Ok(stage_name) => stage_name,
+            Err(_) => {
+                warning!(String::from("Your Maa version might out of date"));
```

Comment:
Log macros accept any type implemented `Display` trait, so there is no need to create a `String` from `&str`.